R/asis-nodes.R

```diff
@@ -0,0 +1,199 @@
+find_inline_math <- function(body, ns) {
+   i <- ".//md:text[not(@asis) and contains(text(), '$') and not(contains(text(), '$$'))]"
+   xml2::xml_find_all(body, i, ns = ns)
+}
+
+# Helper function to return the proper regex for inline math.
+# Having the start and stop type individually allows me to invert the
+# union between them to find the incomplete cases.
+inline_dollars_regex <- function(type = c("start", "stop", "full")) {
+  # any space
+  ace   <- "[:space:]"
+  punks <- glue::glue("[{ace}[:punct:]]")
+  # Note about this regex: the first part is a lookahead (?=...) that searches 
+  # for the line start, space, or punctuation. Importantly about lookaheads,
+  # they do not consume the string 
+  # (https://junli.netlify.app/en/overlapping-regular-expression-in-python/)
+  # 
+  # The rest of the regex looks for a dollar sign that does not butt up against
+  # a space or another dollar. 
+  start <- glue::glue("(?=^|{punks})[$]?[$][^{ace}$]")
+  stop  <- glue::glue("[^{ace}$][$][$]?(?={punks}|$)")
+  switch(type,
+    start = start,
+    stop = stop,
+    full = glue::glue('({start}.+?{stop})')
+  )
+}
+
+# Find incomplete cases for inline math
+find_broken_math <- function(math) {
+  txt <- xml2::xml_text(math)
+  start <- grepl(inline_dollars_regex("start"), txt, perl = TRUE)
+  stop  <- grepl(inline_dollars_regex("stop"), txt, perl = TRUE)
+  incomplete <- !(start & stop)
+  list(
+    no_end = start & incomplete, 
+    no_beginning = stop & incomplete
+  )
+}
+
+#' Find and protect all inline math elements
+#' 
+#' @param body an XML document
+#' @param ns an XML namespace
+#' @return a modified _copy_ of the original XML document
+#' @keywords internal
+#' @examples
+#' txt <- commonmark::markdown_xml(
+#'   r"{This sentence contains $I_A$ $\frac{\pi}{2}$ inline $\LaTeX$ math.}"
+#' )
+#' txt <- xml2::read_xml(txt)
+#' cat(to_md(list(body = txt, yaml = "")), sep = "\n")
+#' ns  <- xml2::xml_ns_rename(xml2::xml_ns(txt), d1 = "md")
+#' protxt <- tinkr:::protect_inline_math(txt, ns)
+#' cat(to_md(list(body = protxt, yaml = "")), sep = "\n")
+protect_inline_math <- function(body, ns) {
+  body  <- copy_xml(body)
+  math  <- find_inline_math(body, ns)
+  if (length(math) == 0) {
+    return(body)
+  }
+  broke <- find_broken_math(math)
+
+  bespoke  <- !(broke$no_end | broke$no_beginning)
+  endless  <- broke$no_end[!bespoke]
+  headless <- broke$no_beginning[!bespoke]
+
+  imath   <- math[bespoke]
+  bmath   <- math[!bespoke]
+
+  # protect math that is strictly inline
+  if (length(imath)) {
+    new_nodes <- lapply(fix_fully_inline(imath), xml2::xml_children)
+
+    # since we split up the nodes, we have to do this node by node
+    for (i in seq(new_nodes)) {
+      add_node_siblings(math[bespoke][[i]], new_nodes[[i]], remove = TRUE)
+    }
+  }
+
+  # protect math that is broken across lines
+  if (length(bmath)) {
+    if (length(bmath[endless]) != length(bmath[headless])) {
+      stop("Uneven math elements")
```

Comment:
codecov tells me there's no test for it, would it make sense to add one?

Reply:
That's a good point. For context, this is here to make sure the number of opening delimiters match the closing delimiters when LaTeX math splits across two lines or contains markdown elements (e.g. `_i` gets converted into `<emph>` if there is a following `_`).

The regex I wrote (`((?=^|[[:space:][:punct:]])[$]?[$][^[:space:]$].+?[^[:space:]$][$][$]?(?=[[:space:][:punct:]]|$))`) has lookaheads to make sure that things like `$2 and $6 are stored in object$thing` don't match and It's difficult to think of a situation where this would be triggered.... unless someone were documenting their BASIC code that parses financial questions without using a code block:

From the first example of: http://qb64.org/wiki/INKEY$

```markdown
Here's a $2 program:

PRINT "Press any keyboard typing key to end SLEEP"
SLEEP
DO: K$ = INKEY$: PRINT K$: LOOP UNTIL K$ = ""
```

Reply:
I've updated this and added an example in 46e9efaad8af1278ae1f6874d20252c73029719